File: typescript/src/compute/value.ts
/**
 * Value model for the `compute` declarative computation layer.
 *
 * Exactly three value types (spec §5): `int` (signed 64-bit, held as BigInt so
 * the semantics match the future Python/Java ports), `str` (Unicode string), and
 * `list` (ordered sequence of int|str). There is no boolean and no float.
 *
 * Coercion rules (spec §8):
 *   - Arithmetic coerces a *1-character digit* string to int; a multi-character
 *     or non-numeric string in an arithmetic slot is an error (use <to_number>).
 *   - String contexts (<join>, <concat>, <pad>, final output) coerce an int to
 *     its decimal form; a whole list is never a string.
 *
 * All runtime failures throw `ComputeError`; static/structural problems are
 * caught earlier by the validator (`checkCompute`).
 */

export type Value =
  | { readonly t: 'int'; readonly v: bigint }
  | { readonly t: 'str'; readonly v: string }
  | { readonly t: 'list'; readonly v: readonly Value[] };

/** Runtime error raised while evaluating a compute tree. */
export class ComputeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ComputeError';
  }
}

// Signed 64-bit bounds. Kept explicit so overflow is a deterministic error in
// every language implementation rather than a silent wrap.
export const INT64_MIN = -(2n ** 63n);
export const INT64_MAX = 2n ** 63n - 1n;

/** Wrap a BigInt as an `int` value, guarding the signed 64-bit range. */
export function int(v: bigint): Value {
  return { t: 'int', v: guard64(v) };
}

/** Wrap a string as a `str` value. */
export function str(v: string): Value {
  return { t: 'str', v };
}

/** Wrap an array as a `list` value. */
export function list(v: readonly Value[]): Value {
  return { t: 'list', v };
}

/** Throw if `v` leaves the signed 64-bit range; otherwise return it unchanged. */
export function guard64(v: bigint): bigint {
  if (v < INT64_MIN || v > INT64_MAX) {
    throw new ComputeError(`integer overflow: ${v.toString()} is outside the signed 64-bit range`);
  }
  return v;
}

/**
 * Coerce a value to `bigint` for arithmetic. An int passes through; a
 * *single* digit character coerces; anything else is an error — a multi-digit
 * string must be converted explicitly with <to_number>, which keeps the intent
 * unambiguous.
 */
export function coerceInt(value: Value, context = 'arithmetic'): bigint {
  if (value.t === 'int') return value.v;
  if (value.t === 'str') {
    if (/^[0-9]$/.test(value.v)) return BigInt(value.v);
    const hint = /^-?[0-9]+$/.test(value.v)
      ? ' — wrap it in <to_number> to convert a multi-digit string'
      : '';
    throw new ComputeError(`expected an integer in ${context}, got the string "${value.v}"${hint}`);
  }
  throw new ComputeError(`expected an integer in ${context}, got a list`);
}

/** Coerce an int or str to its string form. A list is never a string. */
export function coerceStr(value: Value): string {
  if (value.t === 'str') return value.v;
  if (value.t === 'int') return value.v.toString();
  throw new ComputeError('cannot use a list where a string is expected');
}

/** Final compute output: an int or str renders to text; a list is an error. */
export function valueToOutput(value: Value): string {
  if (value.t === 'list') {
    throw new ComputeError('compute result must be an int or str, not a list');
  }
  return coerceStr(value);
}

/** Parse a `<to_number>` string: all decimal digits, optional leading `-`. */
export function parseIntStrict(s: string): bigint {
  if (!/^-?[0-9]+$/.test(s)) {
    throw new ComputeError(`<to_number>: "${s}" is not a valid integer`);
  }
  return guard64(BigInt(s));
}

/**
 * Euclidean remainder: always non-negative, in `[0, |b|)`, independent of the
 * host language's native `%` sign convention (spec §7.4).
 */
export function euclideanMod(a: bigint, b: bigint): bigint {
  if (b === 0n) throw new ComputeError('<mod>: the modulus (second child) must not be zero');
  const abs = b < 0n ? -b : b;
  const r = a % abs;
  return r < 0n ? r + abs : r;
}

/** Floor (toward −∞) integer division (spec §7.4). */
export function floorDiv(a: bigint, b: bigint): bigint {
  if (b === 0n) throw new ComputeError('<divide>: the divisor (second child) must not be zero');
  let q = a / b;
  if (a % b !== 0n && a < 0n !== b < 0n) q -= 1n;
  return q;
}

File: rust/src/compute/value.rs
//! The value model of the compute layer: three types, and no more.
//!
//! An integer, a string, or a list of those. No boolean and no floating point —
//! deliberately. Every check digit in the world is integer arithmetic over the
//! characters of a string, and a float in that computation is a way to get the
//! wrong answer on one number in a million and never find out which.
//!
//! The reference holds integers in an arbitrary-precision type and range-checks
//! them to signed 64 bits. Rust has no big integer in its standard library, so
//! they are held in `i128` instead — which is exactly enough: the guard rejects
//! anything outside `i64`, and the only way to exceed `i128` before reaching the
//! guard would be to multiply two values that are themselves already out of
//! range. Every arithmetic step is checked, so that cannot happen.

/// Anything the compute layer refuses to do, with the reason it refused.
#[derive(Clone, Debug)]
pub struct ComputeError {
    pub message: String,
}

impl std::fmt::Display for ComputeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for ComputeError {}

pub type ComputeResult<T> = Result<T, ComputeError>;

pub fn err<T>(message: impl Into<String>) -> ComputeResult<T> {
    Err(ComputeError {
        message: message.into(),
    })
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Value {
    Int(i128),
    Str(String),
    Lst(Vec<Value>),
}

impl Value {
    pub fn int(v: i128) -> ComputeResult<Value> {
        Ok(Value::Int(guard64(v)?))
    }

    pub fn str(v: impl Into<String>) -> Value {
        Value::Str(v.into())
    }
}

/// `^-?[0-9]+$`
pub fn is_integer_text(s: &str) -> bool {
    let digits = s.strip_prefix('-').unwrap_or(s);
    !digits.is_empty() && digits.bytes().all(|b| b.is_ascii_digit())
}

pub fn guard64(v: i128) -> ComputeResult<i128> {
    if v < i64::MIN as i128 || v > i64::MAX as i128 {
        return err(format!(
            "integer overflow: {v} is outside the signed 64-bit range"
        ));
    }
    Ok(v)
}

/// Coerce to an integer for arithmetic.
///
/// A *single* digit character coerces, because iterating a string yields
/// characters and summing them is the whole point. A multi-digit string does
/// not: `"12"` in an arithmetic slot is far more often a mistake than an
/// intention, so it has to say `<to_number>` out loud.
pub fn as_int(value: &Value, context: &str) -> ComputeResult<i128> {
    match value {
        Value::Int(i) => Ok(*i),
        Value::Str(s) => {
            let mut chars = s.chars();
            if let (Some(c), None) = (chars.next(), chars.next()) {
                if c.is_ascii_digit() {
                    return Ok(i128::from(c as u8 - b'0'));
                }
            }
            let hint = if is_integer_text(s) {
                " — wrap it in <to_number> to convert a multi-digit string"
            } else {
                ""
            };
            err(format!(
                "expected an integer in {context}, got the string \"{s}\"{hint}"
            ))
        }
        Value::Lst(_) => err(format!("expected an integer in {context}, got a list")),
    }
}

/// An int or a string renders to text. A list never does.
pub fn as_str(value: &Value) -> ComputeResult<String> {
    match value {
        Value::Str(s) => Ok(s.clone()),
        Value::Int(i) => Ok(i.to_string()),
        Value::Lst(_) => err("cannot use a list where a string is expected"),
    }
}

pub fn to_output(value: &Value) -> ComputeResult<String> {
    match value {
        Value::Lst(_) => err("compute result must be an int or str, not a list"),
        other => as_str(other),
    }
}

pub fn parse_int_strict(s: &str) -> ComputeResult<i128> {
    if !is_integer_text(s) {
        return err(format!("<to_number>: \"{s}\" is not a valid integer"));
    }
    match s.parse::<i128>() {
        Ok(v) => guard64(v),
        Err(_) => err(format!(
            "integer overflow: {s} is outside the signed 64-bit range"
        )),
    }
}

/// Euclidean remainder — always in `[0, |b|)`.
///
/// Not the host language's `%`: Rust, C#, Java and JavaScript all give a
/// negative remainder for a negative dividend, Python does not, and a check
/// digit computed with the wrong sign convention is wrong only for some inputs.
/// Pinning it here makes every implementation agree.
pub fn modulo(a: i128, b: i128) -> ComputeResult<i128> {
    if b == 0 {
        return err("<mod>: the modulus (second child) must not be zero");
    }
    let m = b.abs();
    let r = a % m;
    Ok(if r < 0 { r + m } else { r })
}

/// Integer division that rounds toward negative infinity.
pub fn floor_div(a: i128, b: i128) -> ComputeResult<i128> {
    if b == 0 {
        return err("<divide>: the divisor (second child) must not be zero");
    }
    let mut q = a / b;
    let r = a % b;
    if r != 0 && ((a < 0) != (b < 0)) {
        q -= 1;
    }
    Ok(q)
}
